### PR TITLE
 mir-opt: a sub-BB of a cleanup BB must also be a cleanup BB in `EarlyOtherwiseBranch`

### LINKED ISSUE #130769: Panic while compiling http-body v0.4.6 with rustc 1.83.0-nightly (7042c269c 2024-09-23)
### Code

Clone [`http-body`](https://github.com/hyperium/http-body), checkout v0.4.6 and compile in release mode:
```shell
$ git clone https://github.com/hyperium/http-body.git
$ cd http-body
$ git checkout v0.4.6
$ cargo build --release
```

### Meta

`rustc --version --verbose`:
```
rustc 1.83.0-nightly (7042c269c 2024-09-23)
binary: rustc
commit-hash: 7042c269c166191cd5d8daf0409890903df7af57
commit-date: 2024-09-23
host: x86_64-unknown-linux-gnu
release: 1.83.0-nightly
LLVM version: 19.1.0
```

### Error output

```
thread 'rustc' panicked at compiler/rustc_mir_transform/src/validate.rs:152:26:
broken MIR in Item(DefId(0:23 ~ http_body[d100]::collect::{impl#1}::poll)) (after phase change to runtime-optimized) at bb22[2]:
Normal edge to BasicBlockData { statements: [StorageDead(_46)], terminator: Some(Terminator { source_info: SourceInfo { span: src/collect.rs:69:13: 69:14 (#0), scope: scope[1] }, kind: goto -> bb41 }), is_cleanup: false } violates unwind invariants (cleanup true -> false)
stack backtrace:
   0:     0x7fe9e4c02f25 - std::backtrace::Backtrace::create::h8d365723131839b0
   1:     0x7fe9e32738a5 - std::backtrace::Backtrace::force_capture::ha8b2c128ac653152
   2:     0x7fe9e2382ea4 - std[25b33fdea409b280]::panicking::update_hook::<alloc[1694921db2598cc8]::boxed::Box<rustc_driver_impl[91319f0e19a4d720]::install_ice_hook::{closure#0}>>::{closure#0}
   3:     0x7fe9e328b018 - std::panicking::rust_panic_with_hook::hfa739d875be2cf67
   4:     0x7fe9e328adea - std::panicking::begin_panic_handler::{{closure}}::h9a9134f6e05b41bd
   5:     0x7fe9e32887d9 - std::sys::backtrace::__rust_end_short_backtrace::h35fa4dfdd9fe9e0c
   6:     0x7fe9e328aaac - rust_begin_unwind
   7:     0x7fe9e0a9e7b0 - core::panicking::panic_fmt::hc3e7c6c5c469a97f
   8:     0x7fe9e10b311a - <rustc_mir_transform[2c59a960358d0625]::validate::Validator as rustc_mir_transform[2c59a960358d0625]::pass_manager::MirPass>::run_pass
   9:     0x7fe9e3a0dc64 - rustc_mir_transform[2c59a960358d0625]::pass_manager::run_passes_inner
  10:     0x7fe9e4403922 - rustc_mir_transform[2c59a960358d0625]::optimized_mir
  11:     0x7fe9e44021db - rustc_query_impl[16add4a89c515f02]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[16add4a89c515f02]::query_impl::optimized_mir::dynamic_query::{closure#2}::{closure#0}, rustc_middle[5962dfe9f3174182]::query::erase::Erased<[u8; 8usize]>>
  12:     0x7fe9e3a31eae - rustc_query_system[cd92c81c02f5c019]::query::plumbing::try_execute_query::<rustc_query_impl[16add4a89c515f02]::DynamicConfig<rustc_query_system[cd92c81c02f5c019]::query::caches::DefIdCache<rustc_middle[5962dfe9f3174182]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[16add4a89c515f02]::plumbing::QueryCtxt, false>
  13:     0x7fe9e3a3145f - rustc_query_impl[16add4a89c515f02]::query_impl::optimized_mir::get_query_non_incr::__rust_end_short_backtrace
  14:     0x7fe9e440e963 - rustc_middle[5962dfe9f3174182]::query::plumbing::query_get_at::<rustc_query_system[cd92c81c02f5c019]::query::caches::DefIdCache<rustc_middle[5962dfe9f3174182]::query::erase::Erased<[u8; 8usize]>>>
  15:     0x7fe9e4416305 - <rustc_metadata[bda71c6b2cfd99ea]::rmeta::encoder::EncodeContext>::encode_crate_root
  16:     0x7fe9e4ab8ad3 - rustc_metadata[bda71c6b2cfd99ea]::rmeta::encoder::encode_metadata
  17:     0x7fe9e4ac0f54 - rustc_metadata[bda71c6b2cfd99ea]::fs::encode_and_write_metadata
  18:     0x7fe9e4abfd43 - <rustc_interface[1294e3f809d58823]::queries::Linker>::codegen_and_build_linker
  19:     0x7fe9e472a34b - rustc_interface[1294e3f809d58823]::interface::run_compiler::<core[1e18fbba9f4dcc19]::result::Result<(), rustc_span[d14b83658cea8543]::ErrorGuaranteed>, rustc_driver_impl[91319f0e19a4d720]::run_compiler::{closure#0}>::{closure#1}
  20:     0x7fe9e47bf210 - std[25b33fdea409b280]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[1294e3f809d58823]::util::run_in_thread_with_globals<rustc_interface[1294e3f809d58823]::util::run_in_thread_pool_with_globals<rustc_interface[1294e3f809d58823]::interface::run_compiler<core[1e18fbba9f4dcc19]::result::Result<(), rustc_span[d14b83658cea8543]::ErrorGuaranteed>, rustc_driver_impl[91319f0e19a4d720]::run_compiler::{closure#0}>::{closure#1}, core[1e18fbba9f4dcc19]::result::Result<(), rustc_span[d14b83658cea8543]::ErrorGuaranteed>>::{closure#0}, core[1e18fbba9f4dcc19]::result::Result<(), rustc_span[d14b83658cea8543]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[1e18fbba9f4dcc19]::result::Result<(), rustc_span[d14b83658cea8543]::ErrorGuaranteed>>
  21:     0x7fe9e47bf87a - <<std[25b33fdea409b280]::thread::Builder>::spawn_unchecked_<rustc_interface[1294e3f809d58823]::util::run_in_thread_with_globals<rustc_interface[1294e3f809d58823]::util::run_in_thread_pool_with_globals<rustc_interface[1294e3f809d58823]::interface::run_compiler<core[1e18fbba9f4dcc19]::result::Result<(), rustc_span[d14b83658cea8543]::ErrorGuaranteed>, rustc_driver_impl[91319f0e19a4d720]::run_compiler::{closure#0}>::{closure#1}, core[1e18fbba9f4dcc19]::result::Result<(), rustc_span[d14b83658cea8543]::ErrorGuaranteed>>::{closure#0}, core[1e18fbba9f4dcc19]::result::Result<(), rustc_span[d14b83658cea8543]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[1e18fbba9f4dcc19]::result::Result<(), rustc_span[d14b83658cea8543]::ErrorGuaranteed>>::{closure#1} as core[1e18fbba9f4dcc19]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  22:     0x7fe9e47bfc6b - std::sys::pal::unix::thread::Thread::new::thread_start::h9c0c387d4b02e16a
  23:     0x7fe9e5fd16d7 - start_thread
  24:     0x7fe9e605560c - __clone3
  25:                0x0 - <unknown>


rustc version: 1.83.0-nightly (7042c269c 2024-09-23)
platform: x86_64-unknown-linux-gnu

query stack during panic:
#0 [optimized_mir] optimizing MIR for `collect::<impl at src/collect.rs:37:1: 37:36>::poll`
end of query stack
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
<backtrace>
```

</p>
</details>

### PR DESCRIPTION
**Pull request by dianqk**
Fixes #130769.

r? @cjgillot or mir-opt

### CODE REVIEW

File: compiler/rustc_middle/src/mir/mod.rs [hunk 1]
```diff
@@ -1337,8 +1337,8 @@ pub struct BasicBlockData<'tcx> {
 }
 
 impl<'tcx> BasicBlockData<'tcx> {
-    pub fn new(terminator: Option<Terminator<'tcx>>) -> BasicBlockData<'tcx> {
-        BasicBlockData { statements: vec![], terminator, is_cleanup: false }
+    pub fn new(terminator: Option<Terminator<'tcx>>, is_cleanup: bool) -> BasicBlockData<'tcx> {
```
Comment (DianQK) [inline compiler/rustc_middle/src/mir/mod.rs:1340]: This should help remind us to set `is_cleanup`.

File: tests/mir-opt/early_otherwise_branch.rs [hunk 2]
```diff
@@ -78,9 +78,57 @@ fn opt4(x: Option2<u32>, y: Option2<u32>) -> u32 {
...
+    }
+}
+
 fn main() {
     opt1(None, Some(0));
     opt2(None, Some(0));
     opt3(Option2::None, Option2::Some(false));
     opt4(Option2::None, Option2::Some(0));
+    opt5(0, 0);
+    opt5_failed(0, 0);
```
Comment (wesleywiser) [inline tests/mir-opt/early_otherwise_branch.rs:133]: Can we also include a version of the reproducer found here as a test case? https://github.com/rust-lang/rust/issues/130769#issuecomment-2370443086

Comment (DianQK) [reply tests/mir-opt/early_otherwise_branch.rs:133]: I have added this: https://github.com/rust-lang/rust/pull/130786/files#diff-50c7b6495288a4628220c4cd9bba0ccd732c336f03738681353d6b64dc0cfa79R29-R43. But, I think the current `fn unwind` test case is sufficient.

**Comment by DianQK** [comment]
:) ping?

**Comment by DianQK** [comment]
@cjgillot The actual changes in this PR are small, so it should be easy to review. Feel free to re-roll if you like.

**Comment by DianQK** [comment]
@cjgillot three week ping :p

File: compiler/rustc_mir_transform/src/early_otherwise_branch.rs [hunk 3]
```diff
@@ -183,15 +183,17 @@ impl<'tcx> crate::MirPass<'tcx> for EarlyOtherwiseBranch {
...
+            let eq_switch = BasicBlockData::new(
+                Some(Terminator {
+                    source_info: bbs[parent].terminator().source_info,
+                    kind: TerminatorKind::SwitchInt {
+                        // switch on the first discriminant, so we can mark the second one as dead
+                        discr: parent_op,
+                        targets: eq_targets,
+                    },
+                }),
+                bbs[parent].is_cleanup,
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/early_otherwise_branch.rs:195]: Should we just skip trying to optimize `parent` if it's a cleanup block?

Comment (DianQK) [reply compiler/rustc_mir_transform/src/early_otherwise_branch.rs:195]: This previously caused a regression, so I have skipped it.

**Comment by DianQK** [comment]
Three weeks ping

r? cjgillot
r? oli-obk

**Review by oli-obk — APPROVED** [review]

**Comment by DianQK** [comment]
I have rebased.

@bors r=oli-obk

**Pull request merged**